### PR TITLE
[react, react-dom] Move React DOM types to @types/react-dom

### PR DESCRIPTION
**Pull request by pshrmn**
So, this one is a doozy. This PR moves all DOM (HTML/SVG) type definitions out of the `@types/react` package and into `@types/react-dom`. It does this two ways:

1. Creates a `ReactDOM` namespace in `react-dom/index.d.ts` for all of the DOM types.
2. Uses an ambient module for `react` to overload the React API functions that take host components.

The reasoning behind this change is explained in #32260.

--------------

This also touches a "few" other packages. Each affected package has the changes made in its own commit. I have a list of the packages and their changes, but I'm not 100% certain that it is exhaustive.

<details>
  <summary>Affected Packages</summary>

| Package | Change | Types | Tests |
|---------|--------|-------|-------|
| aphrodite | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| assert-equal-jsx | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| atlaskit__button | Use `ReactDOM` namespace in types| :heavy_check_mark: | |
| atlaskit__calendar | Use `ReactDOM` namespace in types| :heavy_check_mark: | |
| atlaskit__single-select | Use `ReactDOM` namespace in types| :heavy_check_mark: | |
| braft-editor | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| chai-enzyme | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| enzyme | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| draft-js | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| cleave.js | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| create-subscription | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| emoji-mart | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-beautiful-dnd | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| fixed-data-table | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| fixed-data-table-2 | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| flux | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| gestalt | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| hedron | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| jsnox | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| lingui__macro | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| lingui__react | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| linkifyjs | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| loadable__component | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| loadable__server | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| material-ui-datatables | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| material-ui | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| meteor | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| meteor-universe-i18n | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| mjml-react | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| navigation-react | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| next | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| muicss | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| next-redux-saga | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| next-redux-wrapper | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| next-server | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| ngreact | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| radium | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| rc-progress | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| rc-slider | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| rc-tooltip | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| rc-tree | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-addons-test-utils | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-dom-factories | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-addons-linked-state-mixin | Extend `react-dom` ambient module, not `react` | :heavy_check_mark: | |
| react-addons-transition-group | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-simple-maps | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-sidebar | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-transition-group | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-select | Import types from `react-dom` in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-slick | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-smooth-scrollbar | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-sortable-pane | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-virtualized | Import types from `react-dom` in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-sortable-tree | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-sparklines | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-star-rating-component | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-splitter-layout | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-spinkit | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-sticky-el | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-stickynode | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-sticky | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-stripe-elements | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-svg-inline | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-swipe | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-adal | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-addons-shallow-compare | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-alert | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-alert/v2 | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| reach__router | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-alice-carousel | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-albus | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-app | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-aria-modal | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-aria-menubutton | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-avatar-editor | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-autocomplete | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-autosuggest | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-body-classname | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-bootstrap-date-picker | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-bootstrap | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-bootstrap-daterangepicker | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-big-calendar | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-breadcrumbs | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-bootstrap-typeahead | Reference `@types/react-dom` in tests | | :heavy_check_mark: | 
| react-bootstrap-typeahead | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-broadcast | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-burger-menu | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-bootstrap-table | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-calendar-timeline | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-coinhive | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-circular-progressbar | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-collapse | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-collapsible | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-confirm | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-content-loader | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-copy-to-clipboard | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-countup | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-cropper | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-css-modules | Extend `react-dom` ambient module, not `react` and reference `@types/react-dom` in tests | :heavy_check_mark: | :heavy_check_mark: |
| react-css-modules | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-data-grid/v1 | Reference `@types/react-dom` in types, use `ReactDOM` namespace in test | :heavy_check_mark: | :heavy_check_mark: |
| react-data-grid/v2 | Reference `@types/react-dom` in types, use `ReactDOM` namespace in test | :heavy_check_mark: | :heavy_check_mark: |
| react-data-grid | Reference `@types/react-dom` in types, use `ReactDOM` namespace in test | :heavy_check_mark: | :heavy_check_mark: |
| react-datagrid | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-custom-scrollbars/v3 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-custom-scrollbars | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-date-range | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-datepicker | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-dnd-multi-backend | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-document-meta | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-dragtastic | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-dropzone/v3 | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-dropzone/v2 | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-dropzone | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-dynamic-number | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-draft-wysiwyg | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-easy-chart | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-fa | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-file-input | Reference `@types/react-dom` in types, use `ReactDOM` namespace in test | :heavy_check_mark: | :heavy_check_mark: |
| react-facebook-login-component | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-faux-dom | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-file-reader-input | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-filepond | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-facebook-login | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-flags-select | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-flex | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-form/v1 | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-form | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-fontawesome | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-foundation | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-gateway | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-frontload | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-geosuggest | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-google-login-component | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-google-maps-loader | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-google-places-suggest | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-gravatar | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-grid-layout | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-headroom | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-helmet | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-helmet/v4 | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-highlight-words | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-holder | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-highlighter | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-hot-loader | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-image-fallback | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-icon-base | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-image-gallery | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-image-crop | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-images | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-imgpro | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-imageloader | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-infinite-scroll-component | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-infinite | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-input-autosize | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-infinite-scroller | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-input-mask/v1 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-input-mask | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-instantsearch-dom | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-instantsearch-core | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-input-calendar | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-intl/v1 | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-intl | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-joyride | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-json-pretty | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-jsonschema-form | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-lazyload | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-jss | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-leaflet | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-lazylog | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-list | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-lifecycle-component | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-loader | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-loadable | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-mailchimp-subscribe | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-map-gl | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-maskedinput | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-mentions | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-measure | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-mdl | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-modal | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-mixin | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-motion | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-motion-slider | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-motion-ui-pack | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-navigation/v2 | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-navigation | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-notification-system | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-no-ssr | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-numeric-input | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-outside-click-handler | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-onclickoutside/v5 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-onclickoutside | Use `ReactDOM` namespace in types & tests | :heavy_check_mark: | :heavy_check_mark: |
| react-onsenui | Use `ReactDOM` namespace in types & tests | :heavy_check_mark: | :heavy_check_mark: |
| react-overlays | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-panelgroup | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-plotly.js | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-places-autocomplete/v6 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-places-autocomplete | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-pointable | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-popover | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-qr-reader | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-radio-group | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-redux-epic | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-redux-i18n | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-request | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-relay | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-resolver | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-responsive | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-responsive/v1 | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-router/v2 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-router-dom | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-router/v3 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-router | Use `ReactDOM` namespace in tests | | :heavy_check_mark: |
| react-router-bootstrap | Use `ReactDOM` namespace in tests | | :heavy_check_mark: |
| react-s-alert | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-scroll | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-scrollbar | Reference `@types/react-dom` in types | :heavy_check_mark: | |
| react-share | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-select/v1 | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-sortable-hoc | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-swipeable-views | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-swipeable | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-syntax-highlighter | Reference `@types/react-dom` in types, use `ReactDOM` namespace in test | :heavy_check_mark: | :heavy_check_mark: |
| react-svg-pan-zoom | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-tag-autocomplete | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-tagsinput | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-text-mask | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-tether | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-timeago | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-toastr | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-toggle/v2 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-toggle | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| react-tracking | Namespace & reference in tests | | :heavy_check_mark: |
| react-toastify | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-transition-group/v1 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-truncate | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-touch | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-treeview | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-twitter-auth | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-vertical-timeline-component | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-visibility-sensor | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-user-tour | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-webcam | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-window-size | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| react-window | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| react-widgets | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| reactcss | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| reactstrap/v4 | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| rebass__grid | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| recompose | Use `ReactDOM` namespace in tests | | :heavy_check_mark: |
| reactstrap | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| redux-auth-wrapper | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| redux-auth-wrapper/v1 | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| redux-first-router-link | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| redux-form | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| redux-form/v4 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| redux-form/v6 | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| redux-infinite-form | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| reflux | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| rmc-drawer | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| rheostat | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__addon-a11y | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__addon-actions | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__addon-backgrounds | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| rrc | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| slate-react | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| storybook__addon-centered | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__addon-jest | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__addon-links | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| storybook__addon-notes | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__addon-knobs | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__react | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook__react-native | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| storybook-readme | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| styled-jsx | Extend `react-dom` ambient module, not `react` | :heavy_check_mark: | |
| styled-system | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| styled-components | Use `ReactDOM` namespace in tests | | :heavy_check_mark: |
| ui-box | Reference `@types/react-dom` in tests | | :heavy_check_mark: |
| styled-components/v3 | Use `ReactDOM` namespace in tests | | :heavy_check_mark: |
| webscopeio__react-textarea-autocomplete | Use `ReactDOM` namespace in types and tests | :heavy_check_mark: | :heavy_check_mark: |
| victory | Use `ReactDOM` namespace in types | :heavy_check_mark: | |
| styled-react-modal | Use `ReactDOM` namespace in types (INCOMPLETE) | :heavy_check_mark: | |
</details>

--------------

There are also a few packages whose tests are still failing, some of which I may need help resolving.

These failing tests include the v15 versions of React & React DOM. I'm not sure if the DOM types should also be moved between those (I'm leaning towards yes), but that might resolve their problems.

<details>
  <summary>Packages with failing tests</summary>

| Package | Reason |
|---------|--------|
| react-addons-linked-state-mixin | ??? |
| bandagedbd__bdapi | expect error typeof React? |
</details>

--------------

This is obviously a lot, but aside from the commits moving types from `react` to `react-dom`, most are relatively simple changes.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

(if this gets merged, the following should be moved to the top of this comment) 
#### 1. Install `@types/react-dom`

```
npm install @types/react-dom
```

Previously, `@types/react` defined the DOM host components (e.g. `<div>` and `<button>`) that TypeScript uses to validate JSX. These have now been moved to `@types/react-dom`.

If you do not have `@types/react-dom` installed, TypeScript cannot validate DOM host components. This means that the following component could not be validated:

```tsx
import React from "react";
export default function Greeting() {
  return <div>Hi!</div>;
}
```

By installing `@types/react-dom`, DOM JSX will automatically be validated. No imports are required for this validation.

**Note:** This change also means that TypeScript will not falsely validate DOM types in non-DOM applications.

#### 2. Import DOM types from `react-dom`

DOM type definitions have now been moved to `@types/react-dom`. This includes types for  DOM elements (`ReactHTML.div`), DOM element attributes (`AnchorHTMLAttributes`), and DOM events (`SyntheticEvent`).

These types should now be imported from `react-dom` instead of `react`.

```diff
- import { MouseEvent } from "react";
+ import { MouseEvent } from "react-dom";
// or
- import * as React from "react";
- React.MouseEvent
+ import * as ReactDOM from "react-dom";
+ ReactDOM.MouseEvent
```

### CODE REVIEW

**Comment by pshrmn** [comment]
GitHub is cutting off the earliest commits, so here are links to the most relevant files:

[`react-dom/index.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32322/files#diff-2829c42a674909a3eef2eff558c9bdd0)
[`react/index.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32322/files#diff-96b72df8b13a8a590e4f160cbc51f40c)

I can squash the commits for the other packages if it would make it easier to review, but I kept them separate to make it easier to track the changed packaged.

File: types/react/index.d.ts [hunk 1]
```diff
@@ -2690,183 +976,11 @@ declare global {
...
-            stop: React.SVGProps<SVGStopElement>;
-            switch: React.SVGProps<SVGSwitchElement>;
-            symbol: React.SVGProps<SVGSymbolElement>;
-            text: React.SVGProps<SVGTextElement>;
-            textPath: React.SVGProps<SVGTextPathElement>;
-            tspan: React.SVGProps<SVGTSpanElement>;
-            use: React.SVGProps<SVGUseElement>;
-            view: React.SVGProps<SVGViewElement>;
-        }
+        interface IntrinsicElements { }
```
Comment (ferdaber) [inline types/react/index.d.ts]: Do we need both this and lines 983-984? Looks like they might be redundant.

**Review by ferdaber — APPROVED** [review]
Just one nitpick, otherwise the changes look straightforward. This will most likely require a big ol' banner at the top of the PR to guide users whose types will break when they upgrade. Can you add some kind of extra write up at the top of your PR?

**Review by Jessidhia — CHANGES_REQUESTED** [review]
Just marking this as something I definitely want to look carefully at. If we're going to make a breaking change (yay) let's make it the best possible 🎉

Also, I suspect the ReactHTML namespace can be deleted. It was used for the React <=0.14 predefined React.DOM.foo factories. But I'll double check.

**Comment by ferdaber** [comment]
@Jessidhia We could make it a high amount of breakage (if you will), but as it stands this PR's only break, and migration path to fix, is to simply have a user install `@types/react-dom` If we start removing types it'd be a tougher migration. Food for thought.

**Comment by pshrmn** [comment]
I squashed all of the commits that add a `<reference types="react-dom">` to tests, so hopefully reviewing this PR is a little bit more manageable now.

File: types/react/index.d.ts (hunk 1 above)
Comment (pshrmn) [reply types/react/index.d.ts]: Yeah, that was leftover from when I re-organized things before moving them to `react-dom/index.d.ts`. I cleaned that and another couple of things up with a new commit.

**Comment by pshrmn** [comment]
I have a couple more questions:

1. Is there a reason that the tests for all of the addon packages (`react-addons-*`) are in the `react` packages and not their own packages? My test issue with `react-addons-linked-state-mixin` goes away if I move its tests from `react/test/index.ts` to its own package.

2. Should the v15 React/React DOM definitions also be updated? I'm pretty sure that the answer is yes so that they'll work with other updated package definitions, but I want to get some confirmation.

**Comment by ferdaber** [comment]
I can answer number 2, number 1 predates my tenure as a maintainer. For number 2 I don't think you should. It's considered legacy to depend on those versions of the package.

**Comment by pshrmn** [comment]
@ferdaber I agree that breaking changes on a legacy app would be especially annoying. My concern is that if the v15 definitions aren't updated, any other packages that do use the updated types would break.

If the v15 defs are untouched, then I see three possible package situations:

1. Only support React v15
2. Support React v15 & v16
3. Only support React v16

3 is the easiest to fix: make the breaking change and switch to the `ReactDOM` namespace.

1 might not be too bad. A `path` can be added to the package's `tsconfig.json` to point to `react/v15`.

I'm not sure what would be done with 2. The "best" solution would probably be a split: move the current definitions to a version folder that sets the `path` to `react/v15` and make the breaking changes in the package's main folder. That would also require a maintaining two sets of definitions for the same version of a package, which I'm not sure is allowed. Are definition versions supposed to be kept inline with their package's current version?

**Comment by johnnyreilly** [comment]
That's a pretty good summary @Jessidhia - I've written about this hairy topic before: https://blog.johnnyreilly.com/2018/09/ivan-drago-and-definitely-typed.html

The tl;Dr is: no semver but good reasons why. Pinky swear

**Comment by pshrmn** [comment]
Oh dang, I guess it doesn't warn you after the first time. I'll get this re-made soon and try to be a little more proactive about getting input.

**Comment by cdaringe** [comment]
such a great effort! i am unable to properly TS-ify react-blessed w/out this, as DOM-things are squashing common controls (e.g. `textarea` exists in blessed and DOM)

**Pull request closed without merging**